### ISSUE TITLE
DiagonalGate does not raise when diagonal entries do not have modulus 1

### ISSUE THREAD

**Issue by hhyap** [Qiskit/qiskit-terra#3842]
### Information

- **Qiskit Terra version**: 0.12.0
- **Python version**: 3.6.9
- **Operating system**: Linux

### What is the current behavior?

`diag_gate` allows entries whose modulus are not one.

### Steps to reproduce the problem

```
import qiskit
from numpy import exp,pi

backend = qiskit.Aer.get_backend("statevector_simulator")
qc = qiskit.QuantumCircuit(1)
qc.h(0)
qc.diag_gate([exp(1j*pi/8),0],qc.qregs)  # <-- no error here
qiskit.execute(qc,backend).result().to_dict()
```

### What is the expected behavior?

An error should be raised.

### Suggested solutions
Line 59 in https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/extensions/quantum_initializer/diag.py#L59: 
`if not np.abs( np.abs(z) - 1 ) < _EPS:`

**Comment by JiahanBro** [comment]
Hey guys! I would like to give it a try. Just to be sure I understood it correctly, the second entry of the list which is 0 seems to be the problem right? Maybe the if case in Line 59 only checks the first entry, therefore no error is raised?

**Comment by Dinesh-byte-cloud** [comment]
hey can I have a shot at this

**Comment by Cryoris** [comment]
There's an old PR that should fix this, #4375. It has some merge conflicts, maybe you can check with @ewinston if you can brush it up, or create a new one based on that 🙂
